jwst/pipeline/calwebb_spec2.py

```diff
@@ -179,12 +179,24 @@ def process_exposure_product(
         if exp_type in ['NRS_MSASPEC', 'NRS_IFU']:
             input = self.msa_flagging(input)
 
-        # Extract 2D sub-windows for NIRSpec slit and MSA
-        if exp_type in ['NRS_FIXEDSLIT', 'NRS_BRIGHTOBJ', 'NRS_MSASPEC']:
-            input = self.extract_2d(input)
+        # It isn't really necessary to include 'NRC_TSGRISM' in this list,
+        # but it doesn't hurt, and it makes it clear that flat_field
+        # should be done before extract_2d for all WFSS/GRISM data.
+        if exp_type in ['NRC_GRISM', 'NIS_WFSS', 'NRC_TSGRISM']:
+            # Apply flat-field correction
+            input = self.flat_field(input)
 
-        # Apply flat-field correction
-        input = self.flat_field(input)
+            if exp_type in ['NRS_FIXEDSLIT', 'NRS_BRIGHTOBJ', 'NRS_MSASPEC',
```

Comment:
I think the logic here should just be `if exp_type != NRC_TSGRISM`, the NRS modes have already been excluded by the previous if logic and won't make it here.

Reply:
Yes, that's much better; thanks.